ctru-rs/src/services/uds.rs

```diff
@@ -0,0 +1,960 @@
+    /// # }
+    /// ```
+    #[doc(alias = "udsSendTo")]
+    pub fn send_packet(
+        &self,
+        packet: &[u8],
+        to_nodes: u16,
+        channel: u8,
+        flags: SendFlags,
+    ) -> Result<(), Error> {
+        if self.service_status() == ServiceStatus::Disconnected {
+            return Err(Error::NotConnected);
+        }
+
+        if self.context.unwrap().spectator {
+            return Err(Error::Spectator);
+        }
+
+        let code = ResultCode(unsafe {
+            ctru_sys::udsSendTo(
+                to_nodes,
+                channel,
+                flags.bits(),
+                packet.as_ptr().cast(),
+                packet.len(),
+            )
+        });
+
+        if code.0
+            != ctru_sys::MAKERESULT(
+                ctru_sys::RL_STATUS as _,
+                ctru_sys::RS_OUTOFRESOURCE as _,
+                ctru_sys::RM_UDS as _,
+                ctru_sys::RD_BUSY as _,
+            )
+        {
+            code?;
+        }
+
+        Ok(())
+    }
+
+    /// Pull a packet from the network.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if the service is currently neither connected to nor hosting a network.
+    /// See [`Uds::connect_network()`] to connect to a network or [`Uds::create_network()`] to create one.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # let _runner = test_runner::GdbRunner::default();
+    /// # use std::error::Error;
+    /// # fn main() -> Result<(), Box<dyn Error>> {
+    /// #
+    /// use ctru::services::uds::{ConnectionType, SendFlags, Uds};
+    /// let mut uds = Uds::new(None)?;
+    ///
+    /// let networks = uds.scan(b"HBW\x10", None, None)?;
+    /// uds.connect_network(&networks[0], b"udsdemo passphrase c186093cd2652741\0", ConnectionType::Client, 1)?;
+    /// let packet = uds.pull_packet()?;
+    /// #
+    /// # Ok(())
+    /// # }
+    /// ```
+    #[doc(alias = "udsPullPacket")]
+    pub fn pull_packet(&self) -> Result<Option<(Vec<u8>, u16)>, Error> {
+        if self.service_status() == ServiceStatus::Disconnected {
+            return Err(Error::NotConnected);
+        }
+
+        let mut frame = MaybeUninit::<[u8; Self::RECV_FRAME_SIZE]>::zeroed();
+
+        let mut actual_size = MaybeUninit::uninit();
+        let mut src_node_id = MaybeUninit::uninit();
+
+        ResultCode(unsafe {
+            ctru_sys::udsPullPacket(
+                &self.context.unwrap() as *const _,
+                frame.as_mut_ptr().cast(),
+                Self::RECV_FRAME_SIZE,
+                actual_size.as_mut_ptr(),
+                src_node_id.as_mut_ptr(),
+            )
+        })?;
+
+        let frame = unsafe { frame.assume_init() };
+        let actual_size = unsafe { actual_size.assume_init() };
+        let src_node_id = unsafe { src_node_id.assume_init() };
+
+        Ok(if actual_size == 0 {
+            None
+        } else {
+            Some((frame[..actual_size].to_vec(), src_node_id))
```

Comment:
It might be wiser to `truncate()` and possibly `shrink_to_fit()` the original vector, rather than reallocating.

Reply:
I forgot to check the standard library documentation to see if there's a better way to do this. Whoops.